### PR TITLE
Add terraform

### LINKED ISSUE #8: AWS Cloud Template
There's quite a bit of setup involved: an AWS Lambda Function, an IAM role, 2 CloudWatch triggers, and API Gateway with several routes/resources. Rather than setting this all up by hand, it would be nice to have an AWS Cloud template that folks could use to get up and running quickly (along with clear instructions on how to use it).

As a bonus, might be worth investigating if things could be ported to GCE or Azure and provide configuration for those platforms too.

### PR DESCRIPTION
**Pull request by Furchin**
Fixes #8 

This PR adds terraform configuration for a CloudFront-hosted custom domain on Route53 pointing at an API Gateway integration with Lambda. The three CloudWatch triggers that make the Lambda function execute (daily: check for expired invitations, daily: check for repos to archive, every half hour: check for expired windows) are also created. An IAM role granting the Lambda function access to CloudWatch logs (for logs, duh) and SES (to send email notifications) is also set up.

README is updated with hopefully far simpler installation instructions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,7 +19,21 @@ At the end of the assignment window, simply peruse the candidate's repo and judg
...
+## Installation and Setup
+CodeBnb uses [TerraForm](http://terraform.io) to help manage infrastructure.
+
+1. Create an ACM cert valid for your domain, eg `example.com`. The ACM cert should also support a wildcard for your domain, eg `*.example.com`. (Your Lambda function will be called by an API Gateway located at `https://codebnb.example.com`.)
+1. Pick an S3 bucket to use. Set this name in `package.json` under `config` > `s3bucket`.
+1. Create the S3 bucket you want to use to as the source for the AWS Lambda function.
+  1. `$ npm run createS3bucket`
+1. Change to the terraform directory
+  1. `$ cd terraform`
+1. Install the AWS TerraForm provider.
```
Comment (JaredReisinger) [inline README.md]: pedantic nit: it's "Terraform", not "TerraForm".

File: README.md [hunk 2]
```diff
@@ -19,7 +19,21 @@ At the end of the assignment window, simply peruse the candidate's repo and judg
...
+  1. `$ terraform init`
+1. Deploy everything using TerraForm. `package.json` contains a convenience script which (a) bundles everything into a zip file, (b) uploads it to a version folder in the S3 bucket, (c) sets the s3 bucket and version variables in TerraForm, and (d) runs `terraform apply` for you.
+  1. `$ npm run deploy` and run the `terraform apply` command instructed. You can manually modify the AWS Lambda environment variables afterwards.
```
Comment (JaredReisinger) [inline README.md]: The parent point says that the convenience script does everything, including "(d) runs `terraform apply` for you", but the child bullet says you will be instructed to run `terraform apply` by the script.  One of these is likely wrong!

Comment (JaredReisinger) [reply README.md]: Looking at `package.json`, the parent point is misleading... it should be "(d) shows you the exact `terraform apply` command you need to run." ... but I'm not sure why it _doesn't_ just run `terraform apply` on your behalf, unless there's a good chance you might want/need to alter the command slightly, or perform other interim steps.

File: package.json [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
+  "config": {
+    "s3bucket": "codebnb-sample-bucket",
+    "awsRegion": "us-east-1"
+  },
   "scripts": {
     "test": "jest --coverage",
-    "zip": "zip -r lambda.zip *.js package.json node_modules html"
+    "zip": "zip -r lambda.zip *.js package.json node_modules html",
+    "createS3bucket": "aws s3api create-bucket --bucket $npm_package_config_s3bucket --region=$npm_package_config_awsRegion",
+    "deploy": "npm version patch && $npm_package_scripts_zip && aws s3 cp lambda.zip s3://$npm_package_config_s3bucket/v$npm_package_version/lambda.zip && echo Interactive variable input required. Please execute: terraform apply -var=\"aws_region=$npm_package_config_awsRegion\" -var=\"app_version=$npm_package_version\" -var=\"s3_bucket=$npm_package_config_s3bucket\""
```
Comment (JaredReisinger) [inline package.json]: I still think `npm run zip` is more explicit than `$npm_package_scripts_zip`... the latter requires you to understand how nested `package.json` values are exposed to the scripts, and that the `zip` script value is simply getting expanded/inlined into the current command.  The former is more explicit: "run the zip script!"

File: terraform/codebnb.tf [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    Name = "${var.lambda_function_name}"
+    Environment = "production"
+  }
+}
+
+provider "aws" {
+  region = "${var.aws_region}"
+}
+
+# TODO: Verify if we can replace references to ${data.aws_region.current.name} with ${var.aws_region}
```
Comment (JaredReisinger) [inline terraform/codebnb.tf]: I believe the answer is "yes".

File: terraform/codebnb.tf [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+  action        = "lambda:InvokeFunction"
+  function_name = "${aws_lambda_function.code_bnb.arn}"
+  principal     = "apigateway.amazonaws.com"
+
+  # More: http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html
+  source_arn = "arn:aws:execute-api:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:${aws_api_gateway_rest_api.code_bnb.id}/*/${aws_api_gateway_method.get_assignment.http_method}${aws_api_gateway_resource.assignment.path}"
+  #source_arn = "${aws_api_gateway_rest_api.code_bnb.execution_arn}/*/*"
+}
+
+# TODO: Figure out what demo is
```
Comment (JaredReisinger) [inline terraform/codebnb.tf]: ???  You never refer to the name of this particular resource, so you can call it whatever you want.  Or is your question: "Figure out why we're provisioning an API gateway account in the first place"?

File: README.md (hunk 2 above)
Comment (Furchin) [reply README.md]: Fixed up the readme ...  the reason it doesn't run it for you is I encountered issues where terraform failed to properly collect user input interactively from me from variables and complained about missing values. Not 100% sure this wasn't a symptom of another problem though, but for now I'll leave it :)

File: package.json (hunk 3 above)
Comment (Furchin) [reply package.json]: Lots of things require people to understand `$npm_package_X` :)

File: terraform/codebnb.tf (hunk 5 above)
Comment (Furchin) [reply terraform/codebnb.tf]: Both, I guess. (originally intended to question the name of the resource, but now yeah unclear what an API gateway account is to me). Leaving as a TODO, maybe someone will contribute a patch -- I know this applied successfully to my AWS account but not 100% confident that if I remove and re-install without the API gateway account that there won't be some pre-existing state as a result. If CodeBnb gets adoption we'll let others help with their experiences applying.

**Pull request merged**